### PR TITLE
Implemented stand alone speech to text component

### PR DESCRIPTION
**Pull request by groogiam**
radzenhq#699 

Refactored original pull request: radzenhq#700

### CODE REVIEW

**Comment by akorchev** [comment]
Thank you for making a new PR! I have added a few comments to the code that I think needs more attention.

**Comment by groogiam** [comment]
@akorchev I'm not seeing any additional comments.

**Comment by groogiam** [comment]
Yep, I see them. Thanks.

File: Radzen.Blazor/themes/components/blazor/_speech-to-text.scss [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    0% {
+    }
+
+    100% {
+        background-color: $rz-primary;
+    }
+}
+
+.rz-mic-on {
+    -webkit-animation: mic-blink 2s linear infinite;
```
[the remark this answers is not in the document]
Comment (groogiam) [reply Radzen.Blazor/themes/components/blazor/_speech-to-text.scss:20]: I have removed them and recompiled the SASS but it they do not seem to be included in the generated output.

This
```scss
@keyframes mic-blink {
    0% {
    }

    100% {
        background-color: $rz-primary;
    }
}

.rz-mic-on {
    animation: mic-blink 2s linear infinite;
}

```

Generates this

```css
@keyframes mic-blink {
  0% {
  }
  100% {
    background-color: #ff6d41;
  }
}

.rz-mic-on {
  animation: mic-blink 2s linear infinite;
}
```

Is there something else I need to add in the scss file to get them to generate?

File: Radzen.Blazor/RadzenSpeechToText.razor [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿@inherits RadzenComponent
+@if (Visible)
+{
+    <RadzenButton @ref="_micButton" ButtonStyle="ButtonStyle.Light" Icon="mic" class="@GetCssClass()" Click="OnSpeechToTextClicked"
```
[the remark this answers is not in the document]
Comment (groogiam) [reply Radzen.Blazor/RadzenSpeechToText.razor:4]: I believe this is still needed as this is not a generic element so it seems to get a reference to the underlying ElementRef is by first getting a reference to the component. I could potentially warp the RadzenButton in a span or something but I think this muddies up the markup.

[the remark this answers is not in the document]
Comment (groogiam) [reply Radzen.Blazor/RadzenSpeechToText.razor:4]: I can patch the micbutton element through to the base ElementRef using something like this. But that still requires the ref to the RadzenButton.

```csharp
/// <inheritdoc />
        protected override void OnAfterRender(bool firstRender)
        {
            base.OnAfterRender(firstRender);

            if (firstRender)
            {
                Element = _micButton.Element;
            }
        }
```

File: RadzenBlazorDemos/Pages/SpeechToTextPage.razor [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+<EventConsole @ref=@console Class="mt-4" />
+
+@code {
+
+    string value;
+
+    EventConsole console;
+
+    RadzenTextArea textToSpeechInput;
```
[the remark this answers is not in the document]
Comment (groogiam) [reply RadzenBlazorDemos/Pages/SpeechToTextPage.razor]: Removed the @ref and just used a bool to determine if the result should be written to the text area.

**Comment by groogiam** [comment]
@akorchev Just pushed a new commit which should address most of the previous comments. I marked the conversations the commit addresses as resolved. I have also replied to comments I still have questions on. Thanks.

File: Radzen.Blazor/RadzenSpeechToText.razor (hunk 2 above)
[the remark this answers is not in the document]
Comment (akorchev) [reply Radzen.Blazor/RadzenSpeechToText.razor]: Maybe we shouldn't set the title and aria-label attributes (or set them before `@attributes=Attributes` so a developer can override them).

File: RadzenBlazorDemos/Pages/SpeechToTextPage.razor [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    void OnChange(string value, string name)
+    {
+        console.Log($"{name} value changed to {value}");
+    }
+
+    void OnSpeechCaptured(string speechValue, RadzenTextArea textArea)
+    {
+        console.Log($"Speech Captured: {speechValue}");
+
+        if (textArea != null)
```
[the remark this answers is not in the document]
Comment (groogiam) [reply RadzenBlazorDemos/Pages/SpeechToTextPage.razor]: This should have been fixed in the latest commit. I renamed the page to reflect the new component name.

**Comment by akorchev** [comment]
This PR seems fine. We will merge it probably next week as it needs some design love from @yordanov. Thank you @groogiam !

**Comment by akorchev** [comment]
Hi @groogiam,

I tested in multiple browsers but the speech recognition never seems to hit the onresult callback and nothing happens in the demo. I see the browser starts recording (there is a michrophone icon in the addressbar) but when I click the button again nothing happens. Through debugging I found that `radzenRecognition.componentRef` was always null. Could you please check?

**Comment by akorchev** [comment]
I've managed to get some progress by removing 

```
            radzenRecognition.componentRef = null;
            radzenRecognition.elementRef = null;
```

and using `stop()` instead of `abort()`.  Now speech recognition works in Safari and Chrome (and still fails in Edge with the same 'language-not-supported' error.

I've noticed that Chrome triggers 'no-speech' error if I am being silent for some time. I think we should handle the 'error' event. Also I am not sure what to do with setting componentRef to null but if we do it immediately after calling stop() it won't work. 

I think the JavaScript interaction needs more work and should probably be rethought.

**Comment by akorchev** [comment]
Another comment - I think if there is an active recognition we should call its stop method if a user starts a new one. Right now it fires a separate event. I think the logic can be simplified a lot if instead of tracking the component reference we pass it as a parameter to initDictation and also set radzenRecognition to null if there is an error or dictation is stopped. I may refactor it that way at some point if you don't have time.

**Comment by akorchev** [comment]
I am afraid I can't merge it like this if it doesn't work for me.

**Comment by groogiam** [comment]
Can you provide some details from your chrome / edge installations. Version, default language, etc? I'll try and investigate to see why we are seeing different behavior.

**Comment by akorchev** [comment]
I have latest Chrome (installed today). 

image

The language seems to be a problem only for Edge so we can leave it for later.

Desktop Safari also doesn't work as it is (tested on a mac). 

Perhaps I should make it work in Safari and Chrome and push in the PR so you can test. Thing is I can't do it any time soon :(

**Comment by groogiam** [comment]
Can you provide the steps you are using to test and perhaps the dev console output when you receive an error? I've retested this now on multiple Windows 10 and 11 machines with both Chrome and Edge and it still seems to be working. I have not tested on Safari though.  I added an error handler and it appears at least in my case a speech recognition error does not prevent recording of additional speech in continuous mode so it seems like something is actually faulting the object in a way that cannot recovered from.

I also did some more research and stop vs abort. What happens with stop on changing controls is that the recognition is not ended fast enough and it produces an exception that faults the connection to the mic.

```
blazor.server.js:1 
        
       [2022-11-23T23:37:03.262Z] Error: Microsoft.JSInterop.JSException: Failed to execute 'start' on 'SpeechRecognition': recognition has already started.
Error: Failed to execute 'start' on 'SpeechRecognition': recognition has already started.
    at Object.toggleDictation (https://localhost:5001/_content/Radzen.Blazor/Radzen.Blazor.js:1712:27)
   at Microsoft.JSInterop.JSRuntime.InvokeAsync[TValue](Int64 targetInstanceId, String identifier, Object[] args)
   at Microsoft.JSInterop.JSRuntimeExtensions.InvokeVoidAsync(IJSRuntime jsRuntime, String identifier, Object[] args)
   at Radzen.Blazor.RadzenSpeechToTextButton.OnSpeechToTextClicked() in C:\Users\Greg.Horvath\source\repos\radzen-blazor-pr-703\Radzen.Blazor\RadzenSpeechToTextButton.razor.cs:line 68
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Radzen.Blazor.RadzenButton.OnClick(MouseEventArgs args) in C:\Users\Greg.Horvath\source\repos\radzen-blazor-pr-703\Radzen.Blazor\RadzenButton.razor.cs:line 137
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)
```
This can probably be worked around with more code but I'm not sure doing that provides any additional useability.

Feel free to submit some additional changes to make it work on your end and I'll be happy to review. I'd still like to keep working on it from my end if I can find a way to reproduce the issues you are having on chrome based on the information requested above. Thanks for all your help.

**Comment by groogiam** [comment]
I just committed some changes to wire up some basic speech recognition error handling. The logic will reset the state to the non recording mode in the event of an error. I think what might be happening is that there is some issues in the browsers configuration that was faulting the speech recognition and because there was nothing trapping those errors the UI was out of sync. This should at least keep the UI in a consistent state if an error occurs. Let me know if this helps in your testing. It should at least provide some additional logging as to what is going on with the radzenSpeechRecognition object.

**Comment by akorchev** [comment]
I recorded my experience with the latest commit.

Google Chrome fires the error event - it thinks speech recognition has been aborted.
chrome

The same happens in Safari

Then I changed `abort()` to `stop()` and commended out the code which clears the componentRef. Recording works as expected in Google Chrome and Safari.
chrome-stop

**Comment by groogiam** [comment]
Can you step through the toggleDictation method with some screenshots? Its almost like it is being called twice on your initial starting of recording. The abort method should only ever be called on a second click and seems like it is being called on the initial click. The code you are commenting is essentially wiping out the state management and is what makes one button stop recording when another is clicked.

**Comment by akorchev** [comment]
It isn't called twice for sure as I am clicking once (it is visible in the gifs that I have recored - there is a "bubble" displayed when I click). Also as can be seen from the second gif everything works a lot better when `stop()` is used.

I am afraid I don't have any more time to spend on this right now. Will implement a solution that works for me and push it here for you to test.

**Comment by groogiam** [comment]
There are two clicks on the first gif. several seconds apart. It actually looks like it is working correctly in that one, just that your browser is not processing any speech to raise the onresult event. The second click is what results in the abort event which is expected.

**Comment by akorchev** [comment]
No, there are no two clicks check again... I have been testing this for a few hours already. Give me a break :)

Here is one last gif. I push the button, it breaks **once**, I say "Test" and click it again - it calls abort which throws the error event.

chrome

**Comment by groogiam** [comment]
That looks like the correct behavior. It appears your browser is just not processing the speech and raising the event or the second click is happening before that happens. Usually it takes a couple seconds for it to do that. Clicking stop is not what triggers the speech recognition event as we have continuous set. This is to allow for long dictations in multiple different events.

Apologies I'm not trying to give you a hard time it just seems like we are not on the same page on how this is intended to work.

**Comment by akorchev** [comment]
I expect when I press the button a second time to get something out of it. And it works just like that if the `stop()` method is used instead of `abort()`. It seems to wait long enough for processing to happen. Abort stops everything immediately and no results are given. Users will not be content with such functionality. I expect the following workflow:

1. User presses the record button.
2. Browser starts recording.
3. User speaks for how long she desires.
4. User presses the record button again.
5. Component fires Change event with the text that has been recognized so far.

**Comment by groogiam** [comment]
My thought is a user is going to respond to what they see on the screen. The buttons are to start and stop recording. No test to speech is instantaneous and explicitly telling it to start and stop seems like reasonable assumption. Also there is a use case for a very long dictation that happens in multiple segments that the flow you propose above does not support. The user would have to press the button multiple times for that to happen.

 This could probably be made to work with stop but as I said there is a race condition that appears manifests when performing dictation on multiple controls sequentially. This may be able to be handled gracefully with more code but again as a user I usually respond to what I see on the screen when it comes to these types of applications. 

Let me see if I can add some extra logic to more gracefully handle the stop, I think that might get us past this UX issue.

P.S. Here is a recording of my testing just for reference.


radzenblazor_703_test

**Comment by akorchev** [comment]
> Also there is a use case for a very long dictation that happens in multiple segments that the flow you propose above does not support. The user would have to press the button multiple times for that to happen.

We can accumulate the results from text to speech as they come and emit the change event when the user stops recording. We can disable another recording if it hasn't ended (there is `end` event which may or not help with this regard).

>This could probably be made to work with stop but as I said there is a race condition that appears manifests when performing dictation on multiple controls sequentially.

This is mandatory I think.

The start record  stop record user interface has been with us for years. People know how to use it.

**Comment by groogiam** [comment]
Doing some initial testing with stop it does not seem to have the desired behavior, at least on chrome. If I'm in the middle of talking and click the button trigger stop vs abort it does not seem to process the partially completed result. Even if I click too quickly after it does not seem to process anything. Unless I'm mistaken the user is going to have to wait for the browser to process the speech before the second stop press. So the only stop appears to add here is that it requires extra code to prevent an exception from restarting the service.

>We can accumulate the results from text to speech as they come and emit the change event when the user stops recording. We can disable another recording if it hasn't ended (there is end event which may or not help with this regard).

This seems like poor UX as the user gets no feedback as to what is being processed as they are talking. Especially on a long dictation.

**Comment by akorchev** [comment]
It seems we are in a general disagreement how this component should work. To me it doesn't work at all in its current state. I can't get any text to display in the example no matter what I do - when I press the stop button nothing happens if the abort function is used. You think this is how it should be for some reason. Well I guess we have reached an impasse and we should probably abort the whole idea. A speech recognition component isn't that simple to do properly.

**Comment by groogiam** [comment]
The way I envisioned this working is that I click the button which starts recording. I start talking and seeing speech dictated near Realtime. When I see what I want (or close to what I want on the screen), I click the button a second time which stops processing immediately. If another record button was clicked before the stop then the record on the first button will be immediately be stopped and recording on the other button will be immediately started. This is aligned with my experience with speech dictation in other apps and on my phone. This is how the component is currently working based on my testing on multiple machines and browsers.

There is some disagreement between how results should be aggregated, but the issue we are facing seems be environmental. From what I saw in the GIF's the first click was not triggering abort and was working properly, its just the audio was never processed by the browser, raising the result event before stop was clicked.

If you do not think this a helpful addition to your product or do not like the implementation then feel free to update to how you think it should work close the PR. I'll proceed with integrating the current component into my code base. If you do refactor this or come back to this at some point in the future I'd be happy to help test your implementation.

I appreciate all the time and consideration you put into reviewing and testing this.

**Comment by akorchev** [comment]
> There is some disagreement between how results should be aggregated, but the issue we are facing seems be environmental.

Not really - I am fine as long as the Change event fires at all. The developer can choose to aggregate results or reset the current value.

What bothers me is that right now it doesn't fire the Change event if the user clicks the record button before the first 'onresult' event has fired. This looks to be solvable from my tests by using `stop()` instead of `abort()`. This is the main thing I have a problem with. When I find the time I will push some changes (I may have said this a few times before).

**Comment by akorchev** [comment]
I found time yesterday evening and pushed some changes. Seems to work in Chrome and Safari (Edge still fails with error but it also fails on every page that uses the speech API so it is probably some installation issue).

**Comment by groogiam** [comment]
Thanks for taking time out of your evening. I did some testing and there seems to be some logic errors in the javascript.

The speech recognition is not actually stopped when clicking the button a second time.
1. Click button 1.
2. Speak some text and wait for it to register.
3. Click the button again.
4. The ui indicates the mic is off but if you keep speaking the component will keep processing speech and firing the change event.

Cannot start a recording on the same button twice.
1. Click button one.
2. Speak some text and wait for it to process.
3. Click button one again to stop recording.
4. Click button one again to start recording again.
5. The user interface will not update to indicate recording is happening.

**Comment by akorchev** [comment]
Good catch. I pushed a new change which should hopefully address both.

**Comment by groogiam** [comment]
I did some more testing and it looks like everything is working as expected. Nice work!

**Comment by yordanov** [comment]
I've updated the styles. One usability suggestion - maybe it is a good idea to let people define `StopIcon` or `ActiveIcon` (as well as Title respectively) in addition to Icon:
image

**Pull request merged**